### PR TITLE
luajit: bump new version

### LINKED ISSUE #4614: Use after free in cord_start and obuf_dup
Tarantool version: 1.10, probably newer too.
OS version: Mac, 10.14.4, Mojave
Bug description:
I built tarantool with enabled ASAN and run box/function1.test.lua.
Here is what ASAN detects:
```
Starting instance box...
Run console at unix/:/Users/gerold/Work/Repositories/tarantool/test/var/001_box/box.control
started
mkdir /Users/gerold/Work/Repositories/tarantool/test/var/001_box/box
2019-11-06 16:03:08.381 [73415] main/101/box C> Tarantool 1.10.4-27-ge9078b55b
2019-11-06 16:03:08.382 [73415] main/101/box C> log level 5
2019-11-06 16:03:08.382 [73415] main/101/box I> mapping 117440512 bytes for memtx tuple arena...
2019-11-06 16:03:08.382 [73415] main/101/box I> mapping 134217728 bytes for vinyl tuple arena...
2019-11-06 16:03:08.391 [73415] main/101/box I> instance uuid 6dd01a50-8319-4b70-8db9-009f2474d3cf
2019-11-06 16:03:08.391 [73415] iproto/101/main I> binary: bound to unix/:/Users/gerold/Work/Repositories/tarantool/test/var/001_box/box.socket-iproto
2019-11-06 16:03:08.391 [73415] main/101/box I> initializing an empty data directory
2019-11-06 16:03:08.420 [73415] main/101/box I> assigned id 1 to replica 6dd01a50-8319-4b70-8db9-009f2474d3cf
2019-11-06 16:03:08.420 [73415] main/101/box I> cluster uuid 896a221c-47f2-400f-a5cb-3542d785163b
2019-11-06 16:03:08.423 [73415] snapshot/101/main I> saving snapshot `/Users/gerold/Work/Repositories/tarantool/test/var/001_box/box/00000000000000000000.snap.inprogress'
2019-11-06 16:03:08.425 [73415] snapshot/101/main I> done
2019-11-06 16:03:08.425 [73415] main/101/box I> ready to accept requests
2019-11-06 16:03:08.426 [73415] main/108/checkpoint_daemon I> started
2019-11-06 16:03:08.426 [73415] main/108/checkpoint_daemon I> scheduled the next snapshot at Wed Nov  6 17:38:51 2019
2019-11-06 16:03:08.427 [73415] main/113/console/unix/:/Users/gerold/Work I> started
2019-11-06 16:03:08.427 [73415] main C> entering the event loop
2019-11-06 16:03:08.542 [73415] main/105/main I> -- function1 -  called --
=================================================================
==73415==ERROR: AddressSanitizer: stack-use-after-scope on address 0x000114eff420 at pc 0x000110b9fe12 bp 0x00011c4fe360 sp 0x00011c4fdb10
READ of size 3 at 0x000114eff420 thread T3
    #0 0x110b9fe11 in __asan_memcpy (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x5ee11)
    #1 0x10e3b0171 in obuf_dup obuf.c:161
    #2 0x10e1b8805 in xlog_write_row xlog.c:1324
    #3 0x10db67a1c in xlog_write_entry wal.c:216
    #4 0x10db66558 in wal_write_to_disk wal.c:828
    #5 0x10dc16e25 in cmsg_deliver cbus.c:353
    #6 0x10dc199e1 in cbus_process cbus.c:635
    #7 0x10dc19fe4 in cbus_loop cbus.c:642
    #8 0x10db60612 in wal_writer_f wal.c:897
    #9 0x10d865549 in fiber_cxx_invoke(int (*)(__va_list_tag*), __va_list_tag*) fiber.h:666
    #10 0x10dc0275f in fiber_loop fiber.c:694
    #11 0x10e3ec92d in coro_init coro.c:110

Address 0x000114eff420 is a wild pointer.
SUMMARY: AddressSanitizer: stack-use-after-scope (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x5ee11) in __asan_memcpy
Shadow bytes around the buggy address:
  0x1000229dfe30: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000229dfe40: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000229dfe50: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000229dfe60: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000229dfe70: f1 f1 f1 f1 00 f2 f2 f2 f8 f8 f2 f2 f8 f2 f2 f2
=>0x1000229dfe80: f8 f2 f2 f2[f8]f8 f3 f3 00 00 00 00 00 00 00 00
  0x1000229dfe90: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000229dfea0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000229dfeb0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000229dfec0: 00 00 00 00 00 00 00 00 f1 f1 f1 f1 00 f3 f3 f3
  0x1000229dfed0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
Thread T3 created by T0 here:
    #0 0x110b9a78d in wrap_pthread_create (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x5978d)
    #1 0x10dc05c4e in cord_start fiber.c:1173
    #2 0x10dc0818d in cord_costart fiber.c:1363
    #3 0x10db5fe8a in wal_init wal.c:412
    #4 0x10dafd95d in box_cfg_xc() box.cc:2116
    #5 0x10dafd52c in box_cfg box.cc:2201
    #6 0x10d862f76 in load_cfg main.cc:524
    #7 0x10db73100 in lbox_cfg_load(lua_State*) cfg.cc:61
    #8 0x10dca794c in lj_BC_FUNCC (tarantool:x86_64+0x10044a94c)

==73415==ABORTING
ok - function1
```
This is where the test is stopped:
```
c:call('function1.args', { 15 })
---
- [[15, 'hello']]
...
box.schema.func.drop("function1.args")
---
...
box.schema.func.create('function1.multi_inc', {language = "C"})
---
...
box.schema.user.grant('guest', 'execute', 'function', 'function1.multi_inc')
---
...
c:call('function1.multi_inc')
---
- []
...
box.space.test:select{}
---
- []
...
c:call('function1.multi_inc', { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 })
box.space.test:select{}
[Lost current connection]
c:call('function1.multi_inc', { 2, 4, 6, 8, 10 })
[Lost current connection]
box.space.test:select{}
[Lost current connection]
c:call('function1.multi_inc', { 0, 2, 4 })
```
Probably they are false-positive, but it needs to be checked.

### LINKED ISSUE #4630: ffi.new() isn't JITted, when it should be
We observed some strange decimal tuple field behaviour recently: accessing decimal tuple fields by index is slower than doing the same by name.
For example, run the following script: `time tarantool script.lua`

script.lua:
```lua
local decimal = require('decimal')
local num_iters = 1000000

box.cfg{}

if box.space.test ~= nil then box.space.test:drop() end 
s = box.schema.space.create('test', {format = { 
    { name = 'id', type = 'unsigned' },
    { name = 'dec', type = 'decimal' }
}})
s:create_index('pk')

tup = s:insert{0, decimal.new('1234.5678')}

for i = 1,num_iters do
    a = tup[2]
end

print(a)

os.exit()
```
The output will be:
```
1234.5678

real	0m8.155s
user	0m8.054s
sys	0m0.045s
```
However, if you change `a = tup[2]` to `a = tup.dec` in the loop, the output will be
```
1234.5678

real	0m0.697s
user	0m0.640s
sys	0m0.022s
```

Turns out the problem lies in luajit not JITting ffi.new("decimal_t") call.
jit.dump partial output is below
```
0000  . . . . FUNCF    7          ; msgpackffi.lua:497
0001  . . . . UGET     2   0      ; ffi
0002  . . . . TGETS    2   2   0  ; "new"
0003  . . . . KSTR     3   1      ; "decimal_t"
0004  . . . . CALL     2   2   2
0000  . . . . . FUNCC               ; ffi.new
```
...
```
0245 >  fun EQ     0244  ffi.new
0246 >  str EQ     "decimal_t"  "decimal_t"
0247 >  cdt CNEW   +106
0248    p64 ADD    0247  +8
0249    int XSTORE 0248  +0
0250    p64 ADD    0247  +12
0251    int XSTORE 0250  +0
0252    p64 ADD    0247  +16
0253    u8  XSTORE 0252  +0
---- TRACE 23 abort msgpackffi.lua:497 -- NYI: unsupported C type conversion
```
For the record, ffi's [NYI list](http://luajit.org/ext_ffi_semantics.html#status)  states, that the following operations are not JITted:
> Bitfield accesses and initializations.
> Vector operations.
> Table initializers.
> Initialization of nested struct/union types.
> Allocations of variable-length arrays or structs.
> Allocations of C types with a size > 128 bytes or an alignment > 8 bytes.
> Conversions from lightuserdata to void *.
> Pointer differences for element sizes that are not a power of two.
> Calls to C functions with aggregates passed or returned by value.
> Calls to ctype metamethods which are not plain functions.
> ctype __newindex tables and non-string lookups in ctype __index tables.
> tostring() for cdata types.
> Calls to ffi.cdef(), ffi.load() and ffi.metatype().

Here's our 'decimal_t' C declaration:
```c
typedef struct {
    int32_t digits;
    int32_t exponent;
    uint8_t bits;
    uint16_t lsu[13];
    } decimal_t;
```
Looks like it doesn't fall under any of the restrictions listed above, nevertheless FFI refuses to JIT compile
`ffi.new('decimal_t')`

### LINKED ISSUE #5885: ffi.new() strangely drastic perf difference
`ffi.new()` gives sky and earth different perf when used for structures depending on whether they are allocated as an array. For instance, `ffi.new('struct tt_uuid[1]')` is ~600 times faster than `ffi.new('struct tt_uuid')`. The same works for other small structs, such as `struct port_c`. Here is the reproducer:
```Lua
local fiber = require('fiber')
local clock = require('clock').monotonic64
local ffi = require('ffi')

local run_count = 100
local iter_count = 100000

-- Make it look like there are many allocations. To simulate fragmentation.
local gc_burden = {}
for i = 1, 10000 do
    gc_burden[i] = ffi.new('int[10]')
end

local function bench_ffi_uuid_object()
    for i = 1, iter_count do
        local reg = ffi.new('struct tt_uuid')
        reg.time_low = i
    end
end

local function bench_ffi_uuid_array()
    for i = 1, iter_count do
        local reg = ffi.new('struct tt_uuid[1]')
        reg[0].time_low = i
    end
end

local times = {}
for i = 1, run_count do
    collectgarbage('collect')
    local t1 = clock()

    ------------------------- PUT THE BENCH CALLS HERE -------------------------
    bench_ffi_uuid_array()
    ----------------------------------------------------------------------------

    local t2 = clock()
    local duration = t2 - t1
    table.insert(times, duration)
end

table.sort(times)
print(string.format('Median value per iteration: %s ns',
                    tonumber(times[#times / 2]) / iter_count))
```
In the loop replace `bench_ffi_uuid_array` with `bench_ffi_uuid_object` to see the difference. I used Release build on Mac. The object allocation costs ~300ns per. The array allocation costs 0.5ns.

Probably there is an explanation why is it so, and it would be good to find it.

I thought this had something to do with `ffi.metatype()` being used for `struct tt_uuid`, but it appears the array elements also have the methods, just like object-allocated uuid. And it does not explain `struct port_c`, which does not have a "metatype".

### LINKED ISSUE #6548: Backporting LuaJIT upstream commits part 1
This ticket is an umbrella for all commits we planning to back-port before the qurent Q.
All discarded commits will be backported in scope of #7230.

arm64:
* [x] [06cd9fce](https://github.com/LuaJIT/LuaJIT/commit/06cd9fce) ARM64: Fix assembly of HREFK.
* [x] [9143e864](https://github.com/LuaJIT/LuaJIT/commit/9143e864) FFI/ARM64: Fix pass-by-value struct calling conventions.
* [x] [836fb5bb](https://github.com/LuaJIT/LuaJIT/commit/836fb5bb) Reorganize lightuserdata interning code.
* [x] [16d38a4b](https://github.com/LuaJIT/LuaJIT/commit/16d38a4b) Avoid conflict between 64 bit lightuserdata and ITERN key.

general + x86/x64:
* [x] [630ff319](https://github.com/LuaJIT/LuaJIT/commit/630ff319) Add missing LJ_MAX_JSLOTS check.
* [x] [9f0caad0](https://github.com/LuaJIT/LuaJIT/commit/9f0caad0) Fix FOLD rule for strength reduction of widening.
* [x] [f3cf0d6e](https://github.com/LuaJIT/LuaJIT/commit/f3cf0d6e) Give expected results for negative non-base-10 numbers in tonumber().
* [x] [0a9ff94c](https://github.com/LuaJIT/LuaJIT/commit/0a9ff94c) Actually implement maxirconst trace limit.
* [x] [20ac817a](https://github.com/LuaJIT/LuaJIT/commit/20ac817a) Fix bytecode dump unpatching.
* [x] [46a1b268](https://github.com/LuaJIT/LuaJIT/commit/46a1b268) Add stricter check for print() vs. tostring() shortcut.
* [x] [45a7e507](https://github.com/LuaJIT/LuaJIT/commit/45a7e507) Fix tonumber("-0").
* [x] [103c29e6](https://github.com/LuaJIT/LuaJIT/commit/103c29e6) Fix tonumber("-0") in dual-number mode.
* [x] [92fa45f9](https://github.com/LuaJIT/LuaJIT/commit/92fa45f9) Fix interaction between profiler hooks and finalizers.
* [x] [dfa692b7](https://github.com/LuaJIT/LuaJIT/commit/dfa692b7) Fix string.char() recording with no arguments.
* [x] [e613105c](https://github.com/LuaJIT/LuaJIT/commit/e613105c) Fix write barrier for lua_setupvalue() and debug.setupvalue().
* [x] [f5b0fff5](https://github.com/LuaJIT/LuaJIT/commit/f5b0fff5) Fix debug.debug() for non-string errors.
* [x] [ad65934f](https://github.com/LuaJIT/LuaJIT/commit/ad65934f) FFI: Handle zero-fill of struct-of-NYI.
* [x] [95140c50](https://github.com/LuaJIT/LuaJIT/commit/95140c50) Flush and close output file after profiling run.
* [x] [6b082485](https://github.com/LuaJIT/LuaJIT/commit/6b082485) x64/LJ_GC64: Fix fallback case of asm_fuseloadk64().

### PR DESCRIPTION
**Pull request by igormunkin**
* Avoid conflict between 64 bit lightuserdata and ITERN key.
* Reorganize lightuserdata interning code.
* test: fix path storage for non-concatable objects
* ARM64: Fix assembly of HREFK.
* FFI/ARM64: Fix pass-by-value struct calling conventions.
* test: set DYLD_LIBRARY_PATH environment variable
* x64/LJ_GC64: Fix fallback case of asm_fuseloadk64().
* FFI: Handle zero-fill of struct-of-NYI.
* Fix interaction between profiler hooks and finalizers.
* Flush and close output file after profiling run.
* Fix debug.debug() for non-string errors.
* Fix write barrier for lua_setupvalue() and debug.setupvalue().
* Fix FOLD rule for strength reduction of widening.
* Fix bytecode dump unpatching.
* Fix tonumber("-0") in dual-number mode.
* Fix tonumber("-0").
* Give expected results for negative non-base-10 numbers in tonumber().
* Add missing LJ_MAX_JSLOTS check.
* Add stricter check for print() vs. tostring() shortcut.

Closes #6548
Fixes #4614
Fixes #4630
Fixes #5885
Fixes tarantool/tarantool-qa#234
Fixes tarantool/tarantool-qa#235
Follows up #2712

NO_DOC=LuaJIT submodule bump
NO_TEST=LuaJIT submodule bump

### CODE REVIEW

File: changelogs/unreleased/gh-6548-luajit-fixes.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+* Now initialization of zero-filled struct is compiled (gh-4630, gh-5885).
+* Actually implemented `maxirconst` option for tuning JIT compiler.
+* Fixed JIT stack of Lua slots overflow during recording for metamethod calls.
+* Fixed bytecode dump unpatching for JLOOP in up-recursion compiled functions.
+* Fixed FOLD rule for strength reduction of widening in cdata indexing.
+* Fixed `string.char()` recording without arguments.
+* Fixed `print()` behaviour with reloaded default metatable for numbers.
+* `tonumber("-0")` now saves the sign of number for conversion.
+* `tonumber()` now give predictable results for negative non-base-10 numbers.
+* Fixed write barier for `debug.setupvalue()` and `lua_setupvalue()`.
```
Comment (Buristan) [inline changelogs/unreleased/gh-6548-luajit-fixes.md:16]: Typo: s/barier/barrier/

File: changelogs/unreleased/gh-6548-luajit-fixes.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+activity, the following issues have been resolved:
+
+* Fixed emitting for fuse load of constant in GC64 mode (gh-4095, gh-4199, gh-4614).
+* Now initialization of zero-filled struct is compiled (gh-4630, gh-5885).
+* Actually implemented `maxirconst` option for tuning JIT compiler.
+* Fixed JIT stack of Lua slots overflow during recording for metamethod calls.
+* Fixed bytecode dump unpatching for JLOOP in up-recursion compiled functions.
+* Fixed FOLD rule for strength reduction of widening in cdata indexing.
+* Fixed `string.char()` recording without arguments.
+* Fixed `print()` behaviour with reloaded default metatable for numbers.
```
Comment (Buristan) [inline changelogs/unreleased/gh-6548-luajit-fixes.md:13]: Typo: s/reloaded/the reloaded/

**Review by Buristan — APPROVED** [review]
Hi, thanks for the patch!
LGTM, except a few nits in the changelog.

File: changelogs/unreleased/gh-6548-luajit-fixes.md (hunk 1 above)
Comment (igormunkin) [reply changelogs/unreleased/gh-6548-luajit-fixes.md]: Fixed.

File: changelogs/unreleased/gh-6548-luajit-fixes.md (hunk 2 above)
Comment (igormunkin) [reply changelogs/unreleased/gh-6548-luajit-fixes.md]: Fixed.

**Comment by igormunkin** [comment]
`coveralls` [confirms](https://status.coveralls.io/) unexpected maintenance (on June 30), as a result our coverage workflow fails. Anyway, `coveralls` have already reported coverage for this patchset (see the comment above); all force pushes being made later relate only to commit message and changelog tweaks. Merging the patchset.

**Pull request merged**